### PR TITLE
Add docstring to torch.serialization.register_package

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #104046

### CODE REVIEW

File: torch/serialization.py [hunk 1]
```diff
@@ -90,7 +90,34 @@ def _is_zipfile(f) -> bool:
     return read_bytes == local_header_magic_number
 
 
-def register_package(priority, tagger, deserializer):
+def register_package(
+    priority: int,
+    tagger: Callable[[Union[Storage, torch.storage.TypedStorage, torch.storage.UntypedStorage]], Optional[str]],
```
Comment (albanD) [inline torch/serialization.py:95]: Use type alias for the Union.

**Review by albanD — COMMENTED** [review]
Can you add a small example in the doc?

File: torch/serialization.py [hunk 2]
```diff
@@ -90,7 +90,34 @@ def _is_zipfile(f) -> bool:
...
+
+    To override the deserialization behavior for a device in the global registry, one can register a
+    tagger with a higher priority than the existing tagger.
+
+    This function can also be used to register a tagger and deserializer for new devices.
+
+    Args:
+        priority: Indicates the priority associated with the tagger and deserializer, where a lower
+            value indicates higher priority.
+        tagger: Callable that takes in a storage object and returns its tagged device as a string
```
Comment (janeyx99) [inline torch/serialization.py:113]: Just curious, why can these callables return None? What's the use case when that happens?

The docs are pretty clear already, but an example would help solidify understanding.

**Comment by janeyx99** [comment]
also, nice BE :)

File: torch/serialization.py (hunk 2 above)
Comment (mikaylagawarecki) [reply torch/serialization.py:113]: If a given tagger/deserializer returns None, we continue to the next deserializer until something returns something non-None at which point we terminate and return. 

Tried to capture this here, but not sure if it's very clear :b

> Taggers/deserializers are run in the order given by their :attr:`priority` until a tagger/deserializer returns a
    value that is not `None`.

File: torch/serialization.py [hunk 3]
```diff
@@ -90,7 +91,46 @@ def _is_zipfile(f) -> bool:
...
+            or None.
+        deserializer: Callable that takes in storage object and a device string and returns a storage
+            object on the appropriate device or None.
+
+    Returns:
+        `None`
+
+    Example:
+        >>> # xdoctest: +SKIP("omitted definition of _validate_ipu_device")
+        >>> def _ipu_tag(obj):
```
Comment (albanD) [inline torch/serialization.py:124]: nit: make these regular public functions.

File: torch/serialization.py [hunk 4]
```diff
@@ -90,7 +91,46 @@ def _is_zipfile(f) -> bool:
...
+        >>>     if obj.device.type == 'ipu':
+        >>>         return 'ipu'
+        >>> def _ipu_deserialize(obj, location):
+        >>>     if location.startswith('ipu'):
+        >>>         # fn to validate ipu is available and location is valid
+        >>>         device = _validate_ipu_device(location)
```
Comment (albanD) [inline torch/serialization.py:130]: Just do an assert here I think

File: torch/serialization.py [hunk 5]
```diff
@@ -90,7 +91,46 @@ def _is_zipfile(f) -> bool:
...
+        tagger: Callable that takes in a storage object and returns its tagged device as a string
+            or None.
+        deserializer: Callable that takes in storage object and a device string and returns a storage
+            object on the appropriate device or None.
+
+    Returns:
+        `None`
+
+    Example:
+        >>> # xdoctest: +SKIP("omitted definition of _validate_ipu_device")
```
Comment (albanD) [inline torch/serialization.py:123]: You don't need to skip once you remove the call below

**Review by albanD — APPROVED** [review]
Thanks!

**Pull request closed without merging**